packages/flutter/lib/src/widgets/display_feature_sub_screen.dart

```diff
@@ -0,0 +1,319 @@
+/// in global coordinates. An [anchorPoint] with value `Offset.zero` means the
+/// top-left corner of the available screen space. [avoidBouds] are all the
+/// bounds of the [MediaQueryData.displayFeatures].
+///
+/// See also:
+///
+///  * [DisplayFeatureSubScreen] to understand how sub-screens are defined
+class RenderDisplayFeatureSubScreen extends RenderShiftedBox {
+  /// Creates a render object that positions and sizes its child.
+  RenderDisplayFeatureSubScreen({
+    required Iterable<Rect> avoidBounds,
+    required Offset anchorPoint,
+    RenderBox? child,
+  }) : assert(anchorPoint != null),
+        assert(avoidBounds != null),
+        _anchorPoint = anchorPoint,
+        _avoidBounds = avoidBounds,
+        super(child);
+
+  /// The anchor point used to pick the closest sub-screen.
+  Offset get anchorPoint => _anchorPoint;
+  Offset _anchorPoint;
+  set anchorPoint(Offset value) {
+    assert(value != null);
+    if (_anchorPoint == value)
+      return;
+    _anchorPoint = value;
+    markNeedsLayout();
+  }
+
+  /// Areas of the screen that this render object uses to determine where the
+  /// sub-screens are positioned.
+  Iterable<Rect> get avoidBounds => _avoidBounds;
+  Iterable<Rect> _avoidBounds;
+  set avoidBounds(Iterable<Rect> value) {
+    assert(value != null);
+    if (_iterableEquals<Rect>(_avoidBounds, value))
+      return;
+    _avoidBounds = value;
+    markNeedsLayout();
+  }
+
+  /// This render object needs to know its own size and position on the screen
+  /// in order to properly position its child. The global offset is not
+  /// available during layout, since ancestors of this render object are not yet
+  /// done positioning their children. This is why we cache the position
+  /// obtained during the paint phase. When the global offset changes during
+  /// paint, this render object marks itself as needing layout.
+  Offset _lastOffset = Offset.zero;
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  Size computeDryLayout(BoxConstraints constraints) {
+    return constraints.biggest;
+  }
+
+  @override
+  void performLayout() {
+    if (child == null)
+      return;
+    if (avoidBounds.isEmpty) {
+      child!.layout(BoxConstraints.tight(size));
+      final BoxParentData childParentData = child!.parentData! as BoxParentData;
+      childParentData.offset = Offset.zero;
+    } else {
+      final Rect wantedBounds = _lastOffset & size;
+      final Iterable<Rect> subScreens = _subScreensInBounds(wantedBounds, avoidBounds);
+      final Rect closestSubScreen = _closestToAnchorPoint(subScreens, _finiteOffset(anchorPoint));
+
+      child!.layout(BoxConstraints.tight(closestSubScreen.size));
+      final BoxParentData childParentData = child!.parentData! as BoxParentData;
+      childParentData.offset = closestSubScreen.topLeft - _lastOffset;
+    }
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (offset != _lastOffset) {
+      _lastOffset = offset;
```

Comment:
If this RenderObject is positioned via an OffsetLayer (pretty common), `offset` may be zero (or some other value that just reflects the offset in that OffsetLayer). In other words, `offset` does not give the offset of this RenderObject in the global coordinate system, but `wantedBounds` in `performLayout` appears to assume just this.

Reply:
Also: there is no guarantee that paint is called again when this RenderObject moves on screen (and its offset in the global coordinate system changes). It could be wrapped in a RepaintBoundary and as long as the incoming constraints don't change the framework wouldn't repaint or re-layout this object. So this would violate point #2 "Layout its child when the data in point 1 changes." on your list.

Reply:
Thank you for pointing out the exact way this fails. I can now look at specific scenarios while testing.